code/game/objects/obj_defense.dm

```diff
@@ -15,14 +15,43 @@
 		return
 
 	. = damage_amount
-	obj_integrity = max(obj_integrity - damage_amount, 0)
+
+	update_integrity(obj_integrity - damage_amount)
+
 	//BREAKING FIRST
 	if(integrity_failure && obj_integrity <= integrity_failure * max_integrity)
 		obj_break(damage_flag)
+
 	//DESTROYING SECOND
 	if(obj_integrity <= 0)
 		obj_destruction(damage_flag)
 
+/// Proc for recovering obj_integrity. Returns the amount repaired by
+/obj/proc/repair_damage(amount)
+	if(amount <= 0) // We only recover here
+		return
+	var/new_integrity = min(max_integrity, obj_integrity + amount)
+	. = new_integrity - obj_integrity
+
+	update_integrity(new_integrity)
+
+	if(integrity_failure && obj_integrity > integrity_failure * max_integrity)
+		obj_fix()
+
+/// Handles the integrity of an object changing. This must be called instead of changing integrity directly.
+/obj/proc/update_integrity(new_value)
+	SHOULD_NOT_OVERRIDE(TRUE)
+	new_value = max(0, new_value)
+	if(obj_integrity == new_value)
+		return
+	obj_integrity = new_value
```

Comment:
```suggestion
		return
	. = obj_integrity
	obj_integrity = new_value
```
So we can compare the old value to the new one, and thus know the change.

Reply:
While there's not really a problem with adding a return value here, the procs that call this should already be aware of what the integrity is. I set this to not be overridden as off the top of my head I couldn't think of a good reason for a child type to override this instead of one of the other procs that indicate the direction the integrity changed, that can be easily changed in the future if someone implements something it makes sense for.

The code change you suggest here would break things but I assume you meant to return the changed amount not the old value.